compiler: Add option to use Functions, in place of Arrays, for compiler-generated temporary

FabioLuporini opened the pull request.

mloubout commented on devito/types/dense.py:
```diff
@@ -1454,3 +1453,99 @@ def parent(self):
     _pickle_kwargs = Function._pickle_kwargs + ['parent']
+
+
+class TempFunction(DiscreteFunction):
```
Can we have some docstrings?

FabioLuporini replied:
adding

mloubout commented on devito/types/dense.py:
```diff
@@ -1454,3 +1453,99 @@ def parent(self):
+        pointer_dim = kwargs.get('pointer_dim')
+        dimensions = as_tuple(kwargs['dimensions'])
+        if pointer_dim not in dimensions:
+            # This is a bit hacky but it does work around duplicate dimensions when
```
That sounds safe 😓

mloubout commented on devito/types/dense.py:
```diff
@@ -1454,3 +1453,99 @@ def parent(self):
+        if halo is None:
+            halo = tuple((0, 0) for _ in dimensions)
+        if pointer_dim is not None:
+            halo = ((0, 0),) + halo
```
So `TempFunction` can't have spatial order halo (or space_order for that matters)

FabioLuporini replied:
indeed. Irrelevant as the users doesn't create them explicitly

mloubout commented on devito/types/dense.py:
```diff
@@ -1454,3 +1453,99 @@ def parent(self):
+        return self._pointer_dim
+
+    @property
+    def dim(self):
```
why the need for both?

FabioLuporini replied:
Make it usable where we have a PointerArray, as it's `PointerArray.dim`. It'd be cleaner with a Mixin, I admit... I may add one

mloubout commented on devito/types/dense.py:
```diff
@@ -1454,3 +1453,99 @@ def parent(self):
+            raise ValueError("`shape` must contain %d integers, not %d"
+                             % (self.ndim, len(shape)))
+
+        return Function(name=self.name, dtype=self.dtype, dimensions=self.dimensions,
```
What if any of the dimensions is time?

FabioLuporini replied:
undefined behaviour. But temporaries are never created along time, so we should be fine

mloubout replied:
Can we add a check there is no time dim for safety?

FabioLuporini replied:
Done, added a few lines above in `__indices_setup__` , and I've actually made it stronger by checking for `not any(i.is_NonlinearDerived ...)`

mloubout reviewed (commented):
Still not sure what this is for, the test makes it look like it's not really a user side object and very intricate to use (requires to know what the generated code exactly is).

FabioLuporini commented:
@mloubout 

> Still not sure what this is for, the test makes it look like it's not really a user side object and very intricate to use (requires to know what the generated code exactly is).

You're right. I'm going to add docstrings in my next push.

As you correctly guessed, this is _not_ a user API object. These may be created by Devito in place of Arrays to implement tensor temporaries. The key difference is that an Array mallocs and frees memory directly in C-land, while a TempFunction is gonna give control of memory management to the user. I do not expect the average user to care about this (TempFunctions are never going to be used unless the user requests them through an Operator option).

mloubout commented on devito/types/dense.py:
```diff
@@ -1454,3 +1453,133 @@ def parent(self):
+      .. code-block:: python
+
+        op = Operator(...)
+        cfuncs = [i for i in op.input if i.is_TempFunction]
```
May wanna add an `op.tmp_function` property for ease of use.

FabioLuporini replied:
Done, added `op.temporaries`

mloubout reviewed (commented):
Minor comments

tjb900 commented on tests/test_dse.py:
```diff
@@ -1787,6 +1787,67 @@ def test_blocking_options(self, rotate):
+        op0(time_M=1, nthreads=nthreads)
+
+        # TempFunctions expect an override
+        try:
```
Just saw this when looking for this example code. I think pytest gives you a mechanism for asserting that exceptions are thrown (and indeed, as I look again I see that this code _won't_ currently error in the case that no exception is thrown)

```
with pytest.raises(InvalidArgument):
    op1(...)
```

FabioLuporini replied:
ah, good to know! Will update the test.

> as I look again I see that this code won't currently error in the case that no exception is thrown)

That's my mistake. I should have added an exception below `op1(...)`

tjb900 commented on tests/test_dse.py:
```diff
@@ -1787,6 +1787,67 @@ def test_blocking_options(self, rotate):
+            assert False
+
+        # Prepare to run op1
+        shape = [nthreads, x0_blk0_size, y0_blk0_size, grid.shape[-1]]
```
ohhh...  I had not thought about this. How is one supposed to know what to pass to `make`, here - if it's not always the `grid.shape`?

This seems to mean that the caller has to be aware of the details of the generated code. I guess we can live with that, but I'd prefer more safeguards if possible.  Could there potentially be some protection in `arg_values` that ensures the passed temporary is sufficiently large, given the rest of the arguments?

The sequence of events I'm worried about is:
 * devito version X produces C code that needs temporaries of a certain shape
 * application hard-codes temporaries of that shape using a construction very similar to this line
 * devito version X+1 produces C code that needs a slightly larger temporary, but code still passes same size
 * segfaults, incorrect results, chaos reigns
 
Alternatively, or perhaps as well, could the arguments to `make` be _all_ the arguments that will be passed to the operator, and make can extract whatever values it needs from those?

FabioLuporini replied:
>This seems to mean that the caller has to be aware of the details of the generated code.

Well, the idea was that you query the TempFunction's shape. This gives you the _symbolic_ shape of the TempFunction from which you can infer what's required.

> Alternatively, or perhaps as well, could the arguments to make be all the arguments that will be passed to the operator, and make can extract whatever values it needs from those?

yes, if this is feasible on your side, it should be straightforward on our side. So the idea would be (sort of):

```
def make(..., **kwargs):
    # Sanity checks...
  
    try:  
        shape = tuple(i.subs(kwargs) for I in self.shape)
    except ...:
        raise ValueError('Don't know how to derive shape for ...') 
```

?

tjb900 replied:
Ah, yeah I hadn't appreciated that we could get the symbolic shape. That resolves my concern for now, I think.

FabioLuporini commented:
@tjb900 :

* rebased on master
* changed to `with pytest.raises(...)` (thanks!)
* added prototypal API so that you can do `i.make(**kwargs)` instead of providing the shape explicitly. Test updated accordingly
* More roboust sanity checks inside `make`
